File: 60.pl
#!/usr/bin/env perl

# PODNAME: 60.pl
# ABSTRACT: Prime pair sets

## Author     : ian.sealy@sanger.ac.uk
## Maintainer : ian.sealy@sanger.ac.uk
## Created    : 2015-08-07

use warnings;
use strict;
use autodie;
use Getopt::Long;
use Pod::Usage;
use Carp;
use version; our $VERSION = qv('v0.1.0');

# Default options
my $set_size = 5;    ## no critic (ProhibitMagicNumbers)
my ( $debug, $help, $man );

# Get and check command line options
get_and_check_options();

my $limit       = 10;                ## no critic (ProhibitMagicNumbers)
my $min_set_sum = $limit * $limit;

while ( $min_set_sum == $limit * $limit ) {
    $limit *= 10;                    ## no critic (ProhibitMagicNumbers)
    $min_set_sum = $limit * $limit;

    my @primes = get_primes_up_to($limit);

    my %pair;
    foreach my $prime1 (@primes) {
        foreach my $prime2 (@primes) {
            next if $prime2 <= $prime1;
            if (   is_prime( $prime1 . $prime2 )
                && is_prime( $prime2 . $prime1 ) )
            {
                $pair{$prime1}{$prime2} = 1;
            }
        }
    }

    foreach my $prime ( sort { $a <=> $b } keys %pair ) {
        last if $prime > $min_set_sum;
        $min_set_sum = get_set( \%pair, [ keys %{ $pair{$prime} } ],
            [$prime], $set_size, $min_set_sum );
    }
}

printf "%d\n", $min_set_sum;

sub get_set {
    ## no critic (ProhibitReusedNames ProhibitAmbiguousNames)
    my ( $pair, $candidates, $set, $set_size, $min_set_sum ) = @_;
    ## use critic

    my $set_sum = 0;
    foreach my $prime ( @{$set} ) {
        $set_sum += $prime;
    }

    if ( scalar @{$set} == $set_size ) {
        return $set_sum < $min_set_sum ? $set_sum : $min_set_sum;
    }

    foreach my $prime ( sort { $a <=> $b } @{$candidates} ) {
        return $min_set_sum if $set_sum + $prime > $min_set_sum;
        my %intersection = map { $_ => 1 } @{$candidates};
        foreach my $i ( keys %{ $pair->{$prime} } ) {
            $intersection{$i}++;
        }
        my @new_candidates = grep { $intersection{$_} == 2 } keys %intersection;
        $min_set_sum = get_set( $pair, \@new_candidates, [ @{$set}, $prime ],
            $set_size, $min_set_sum );
    }

    return $min_set_sum;
}

sub get_primes_up_to {
    my ($limit) = @_;    ## no critic (ProhibitReusedNames)

    my $sieve_bound = int( ( $limit - 1 ) / 2 );    # Last index of sieve
    my @sieve;
    my $cross_limit = int( ( int( sqrt $limit ) - 1 ) / 2 );
    foreach my $i ( 1 .. $cross_limit ) {
        if ( !$sieve[ $i - 1 ] ) {

            # 2 * $i + 1 is prime, so mark multiples
            my $j = 2 * $i * ( $i + 1 );
            while ( $j <= $sieve_bound ) {
                $sieve[ $j - 1 ] = 1;
                $j += 2 * $i + 1;
            }
        }
    }

    my @primes_up_to = (2);
    foreach my $i ( 1 .. $sieve_bound ) {
        if ( !$sieve[ $i - 1 ] ) {
            push @primes_up_to, 2 * $i + 1;
        }
    }

    return @primes_up_to;
}

sub is_prime {
    my ($num) = @_;

    return 0 if $num == 1;    # 1 isn't prime
    ## no critic (ProhibitMagicNumbers)
    return 1 if $num < 4;         # 2 and 3 are prime
    return 0 if $num % 2 == 0;    # Odd numbers aren't prime
    return 1 if $num < 9;         # 5 and 7 are prime
    return 0 if $num % 3 == 0;    # Numbers divisible by 3 aren't prime
    ## use critic

    my $num_sqrt = int sqrt $num;
    my $factor   = 5;               ## no critic (ProhibitMagicNumbers)
    while ( $factor <= $num_sqrt ) {
        return 0 if $num % $factor == 0;    # Primes greater than three are 6k-1
        return 0 if $num % ( $factor + 2 ) == 0;    # Or 6k+1
        $factor += 6;    ## no critic (ProhibitMagicNumbers)
    }
    return 1;
}

# Get and check command line options
sub get_and_check_options {

    # Get options
    GetOptions(
        'set_size=i' => \$set_size,
        'debug'      => \$debug,
        'help'       => \$help,
        'man'        => \$man,
    ) or pod2usage(2);

    # Documentation
    if ($help) {
        pod2usage(1);
    }
    elsif ($man) {
        pod2usage( -verbose => 2 );
    }

    return;
}

__END__
=pod

=encoding UTF-8

=head1 NAME

60.pl

Prime pair sets

=head1 VERSION

version 0.1.0

=head1 DESCRIPTION

This script solves the Project Euler problem "Prime pair sets". The problem is:
Find the lowest sum for a set of five primes for which any two primes
concatenate to produce another prime.

=head1 EXAMPLES

    perl 60.pl

=head1 USAGE

    60.pl
        [--set_size INT]
        [--debug]
        [--help]
        [--man]

=head1 OPTIONS

=over 8

=item B<--set_size INT>

The target number of primes in the set.

=item B<--debug>

Print debugging information.

=item B<--help>

Print a brief help message and exit.

=item B<--man>

Print this script's manual page and exit.

=back

=head1 DEPENDENCIES

None

=head1 AUTHOR

=over 4

=item *

Ian Sealy <ian.sealy@sanger.ac.uk>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2015 by Genome Research Ltd.

This is free software, licensed under:

  The GNU General Public License, Version 3, June 2007

=cut

File: 60.js
#!/usr/bin/env node

// Prime pair sets

var program = require('commander');

program
    .version('0.1.0')
    .description('Prime pair sets')
    .option('-s, --set_size <int>', 'The target number of primes in the set',
        Number, 5)
    .parse(process.argv);

var limit = 10;
var min_set_sum = limit * limit;

while (min_set_sum == limit * limit) {
    limit *= 10;
    min_set_sum = limit * limit;

    var primes = get_primes_up_to(limit);

    var pair = {};
    for (var i = 0; i < primes.length; i++) {
        for (var j = 0; j < primes.length; j++) {
            if (primes[j] <= primes[i]) {
                continue;
            }
            var pair1 = parseInt(primes[i].toString() + primes[j].toString());
            var pair2 = parseInt(primes[j].toString() + primes[i].toString());
            if (is_prime(pair1) && is_prime(pair2)) {
                if (!pair[primes[i]]) {
                    pair[primes[i]] = {};
                }
                pair[primes[i]][primes[j]] = true;
            }
        }
    }

    primes = Object.keys(pair).map(function(prime) {
        return parseInt(prime);
    }).sort(num_sort);
    for (var i = 0; i < primes.length; i++) {
        if (primes[i] > min_set_sum) {
            break;
        }
        var candidates = Object.keys(pair[primes[i]]).map(function(prime) {
            return parseInt(prime);
        }).sort(num_sort);
        min_set_sum = get_set(pair, candidates, [primes[i]], program.set_size,
            min_set_sum);
    }
}

console.log(min_set_sum);

function get_set(pair, candidates, set, set_size, min_set_sum) {
    var set_sum = 0;
    for (var i = 0; i < set.length; i++) {
        set_sum += set[i];
    }

    if (set.length == set_size) {
        return set_sum < min_set_sum ? set_sum : min_set_sum;
    }

    for (var i = 0; i < candidates.length; i++) {
        if (set_sum + candidates[i] > min_set_sum) {
            return min_set_sum;
        }
        var new_candidates = [];
        if (typeof pair[candidates[i]] != 'undefined') {
            new_candidates = Object.keys(pair[candidates[i]]).sort(num_sort);
        }
        new_candidates = intersection(candidates, new_candidates);
        var new_set = set.slice();
        new_set.push(candidates[i]);
        min_set_sum = get_set(pair, new_candidates, new_set, set_size,
            min_set_sum);
    }

    return min_set_sum;
}

function num_sort(a, b) {
    return a - b;
}

function intersection(a, b) {
    var i = 0;
    var j = 0;
    var intersect = [];
    while (i < a.length && j < b.length) {
        if (a[i] < b[j]) {
            i++;
        } else if (a[i] > b[j]) {
            j++;
        } else {
            intersect.push(a[i]);
            i++;
            j++;
        }
    }

    return intersect;
}

function get_primes_up_to(limit) {
    var sieve_bound = Math.floor((limit - 1) / 2); // Last index of sieve
    var sieve = [];
    var cross_limit = Math.floor((Math.floor(Math.sqrt(limit)) - 1) / 2);
    for (var i = 1; i <= cross_limit; i++) {
        if (!sieve[i - 1]) {
            // 2 * i + 1 is prime, so mark multiples
            var j = 2 * i * (i + 1);
            while (j <= sieve_bound) {
                sieve[j - 1] = true;
                j += 2 * i + 1;
            }
        }
    }

    var primes = [2];
    for (var i = 1; i <= sieve_bound; i++) {
        if (!sieve[i - 1]) {
            primes.push(2 * i + 1);
        }
    }

    return primes;
}

function is_prime(num) {
    if (num == 1) {
        return false; // 1 isn't prime
    } else if (num < 4) {
        return true; // 2 and 3 are prime
    } else if (num % 2 == 0) {
        return false; // Even numbers aren't prime
    } else if (num < 9) {
        return true; // 5 and 7 are prime
    } else if (num % 3 == 0) {
        return false; // Numbers divisible by three aren't prime
    }

    var num_sqrt = Math.floor(Math.sqrt(num));
    var factor = 5;
    while (factor <= num_sqrt) {
        if (num % factor == 0) {
            return false; // Primes greater than 3 are 6k - 1
        } else if (num % (factor + 2) == 0) {
            return false; // Or 6k + 1
        }
        factor += 6;
    }

    return true;
}
